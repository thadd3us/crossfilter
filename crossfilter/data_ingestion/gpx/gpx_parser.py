"""GPX file parsing functions that convert GPX files to DataFrames."""

import hashlib
import logging
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import List, Optional

import gpxpy
import pandas as pd
from gpxpy.gpx import GPXTrackPoint, GPXWaypoint

from crossfilter.core.bucketing import add_geo_h3_bucket_columns
from crossfilter.core.schema import DataType, SchemaColumns as C

logger = logging.getLogger(__name__)


def generate_uuid_from_components(
    first_trackpoint_timestamp: datetime,
    first_trackpoint_lat: float,
    first_trackpoint_lon: float,
    point_timestamp: datetime,
    point_lat: float,
    point_lon: float,
    point_type: DataType,
) -> str:
    """Generate a deterministic UUID from the specified components."""
    # Create a string representation of all components
    components = [
        first_trackpoint_timestamp.isoformat(),
        str(first_trackpoint_lat),
        str(first_trackpoint_lon),
        point_timestamp.isoformat(),
        str(point_lat),
        str(point_lon),
        point_type.value,
    ]

    # Create a deterministic hash from the components
    component_string = "|".join(components)
    hash_digest = hashlib.sha256(component_string.encode("utf-8")).digest()

    # Create a UUID from the hash
    return str(uuid.UUID(bytes=hash_digest[:16]))


def generate_uuid_for_row(
    row: pd.Series,
    first_point_timestamp: datetime,
    first_point_lat: float,
    first_point_lon: float,
) -> str:
    """Generate a UUID for a single DataFrame row."""
    return generate_uuid_from_components(
        first_point_timestamp,
        first_point_lat,
        first_point_lon,
        row[C.TIMESTAMP_UTC].to_pydatetime(),
        row[C.GPS_LATITUDE],
        row[C.GPS_LONGITUDE],
        DataType(row[C.DATA_TYPE]),
    )


def load_gpx_file_to_df(gpx_file_path: Path) -> pd.DataFrame:
    """
    Parse a GPX file and convert it to a DataFrame conforming to the schema.

    Args:
        gpx_file_path: Path to the GPX file

    Returns:
        DataFrame with GPX trackpoints and waypoints

    Raises:
        FileNotFoundError: If the GPX file doesn't exist
        ValueError: If the GPX file is invalid or empty
    """
    if not gpx_file_path.exists():
        raise FileNotFoundError(f"GPX file not found: {gpx_file_path}")

    try:
        with open(gpx_file_path, "r", encoding="utf-8") as f:
            gpx = gpxpy.parse(f)
    except Exception as e:
        raise ValueError(f"Failed to parse GPX file {gpx_file_path}: {e}")

    # Get file size
    file_size = gpx_file_path.stat().st_size

    # Collect all points first (without UUIDs)
    records: List[dict] = []

    # Process all trackpoints
    for track in gpx.tracks:
        for segment in track.segments:
            for point in segment.points:
                if point.time and point.latitude and point.longitude:
                    records.append(
                        {
                            C.DATA_TYPE: DataType.GPX_TRACKPOINT,
                            C.SOURCE_FILE: str(gpx_file_path.name),
                            C.TIMESTAMP_MAYBE_TIMEZONE_AWARE: point.time.isoformat(),
                            C.TIMESTAMP_UTC: (
                                point.time
                                if point.time.tzinfo is not None
                                else point.time.replace(tzinfo=timezone.utc)
                            ),
                            C.GPS_LATITUDE: point.latitude,
                            C.GPS_LONGITUDE: point.longitude,
                        }
                    )

    # Process all waypoints
    for waypoint in gpx.waypoints:
        if waypoint.time and waypoint.latitude and waypoint.longitude:
            records.append(
                {
                    C.DATA_TYPE: DataType.GPX_WAYPOINT,
                    C.NAME: waypoint.name,
                    C.CAPTION: waypoint.comment or waypoint.description,
                    C.SOURCE_FILE: str(gpx_file_path.name),
                    C.TIMESTAMP_MAYBE_TIMEZONE_AWARE: waypoint.time.isoformat(),
                    C.TIMESTAMP_UTC: (
                        waypoint.time
                        if waypoint.time.tzinfo is not None
                        else waypoint.time.replace(tzinfo=timezone.utc)
                    ),
                    C.GPS_LATITUDE: waypoint.latitude,
                    C.GPS_LONGITUDE: waypoint.longitude,
                }
            )

    if not records:
        logger.warning(f"No valid trackpoints or waypoints found in {gpx_file_path}")
        # Return empty DataFrame with correct schema
        df = pd.DataFrame(
            columns=[
                C.UUID_STRING,
                C.DATA_TYPE,
                C.NAME,
                C.SOURCE_FILE,
                C.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
                C.TIMESTAMP_UTC,
                C.GPS_LATITUDE,
                C.GPS_LONGITUDE,
            ]
        )
    else:
        # Create DataFrame
        df = pd.DataFrame(records)

    # Convert TIMESTAMP_UTC to proper datetime with UTC timezone
    df[C.TIMESTAMP_UTC] = pd.to_datetime(df[C.TIMESTAMP_UTC], utc=True)
    df[C.GPS_LATITUDE] = df[C.GPS_LATITUDE].astype(float)
    df[C.GPS_LONGITUDE] = df[C.GPS_LONGITUDE].astype(float)
    df = df.sort_values(
        by=[C.TIMESTAMP_UTC, C.DATA_TYPE, C.GPS_LATITUDE, C.GPS_LONGITUDE]
    )

    if not df.empty:
        # Find the first point for UUID generation (first chronologically)
        first_row = df.iloc[0]
        first_point_timestamp = first_row[C.TIMESTAMP_UTC].to_pydatetime()
        first_point_lat = first_row[C.GPS_LATITUDE]
        first_point_lon = first_row[C.GPS_LONGITUDE]

        # Generate UUIDs for all rows using vectorized operation
        df[C.UUID_STRING] = df.apply(
            lambda row: generate_uuid_for_row(
                row, first_point_timestamp, first_point_lat, first_point_lon
            ),
            axis=1,
        )

    # Set index name
    df.index.name = C.DF_ID

    # Add H3 spatial index columns for each individual GPX file (parallelized computation)
    if not df.empty and C.GPS_LATITUDE in df.columns and C.GPS_LONGITUDE in df.columns:
        add_geo_h3_bucket_columns(df)
        logger.debug(f"Added H3 columns to {len(df)} points from {gpx_file_path}")

    logger.info(f"Loaded {len(df)} points from {gpx_file_path}")

    return df
