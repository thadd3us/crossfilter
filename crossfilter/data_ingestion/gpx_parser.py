"""GPX file parsing functions that convert GPX files to DataFrames."""

import hashlib
import logging
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import List, Optional

import gpxpy
import pandas as pd
from gpxpy.gpx import GPXTrackPoint, GPXWaypoint

from crossfilter.core.schema import DataType, SchemaColumns

logger = logging.getLogger(__name__)


def generate_uuid_from_components(
    first_trackpoint_timestamp: datetime,
    first_trackpoint_lat: float,
    first_trackpoint_lon: float,
    point_timestamp: datetime,
    point_lat: float,
    point_lon: float,
    point_type: DataType,
) -> str:
    """Generate a deterministic UUID from the specified components."""
    # Create a string representation of all components
    components = [
        first_trackpoint_timestamp.isoformat(),
        str(first_trackpoint_lat),
        str(first_trackpoint_lon),
        point_timestamp.isoformat(),
        str(point_lat),
        str(point_lon),
        point_type.value,
    ]
    
    # Create a deterministic hash from the components
    component_string = "|".join(components)
    hash_digest = hashlib.sha256(component_string.encode("utf-8")).digest()
    
    # Create a UUID from the hash
    return str(uuid.UUID(bytes=hash_digest[:16]))


def generate_uuid_for_row(
    row: pd.Series,
    first_point_timestamp: datetime,
    first_point_lat: float,
    first_point_lon: float,
) -> str:
    """Generate a UUID for a single DataFrame row."""
    return generate_uuid_from_components(
        first_point_timestamp,
        first_point_lat,
        first_point_lon,
        row[SchemaColumns.TIMESTAMP_UTC].to_pydatetime().replace(tzinfo=timezone.utc),
        row[SchemaColumns.GPS_LATITUDE],
        row[SchemaColumns.GPS_LONGITUDE],
        DataType(row[SchemaColumns.DATA_TYPE]),
    )


def load_gpx_file_to_df(gpx_file_path: Path) -> pd.DataFrame:
    """
    Parse a GPX file and convert it to a DataFrame conforming to the schema.
    
    Args:
        gpx_file_path: Path to the GPX file
        
    Returns:
        DataFrame with GPX trackpoints and waypoints
        
    Raises:
        FileNotFoundError: If the GPX file doesn't exist
        ValueError: If the GPX file is invalid or empty
    """
    if not gpx_file_path.exists():
        raise FileNotFoundError(f"GPX file not found: {gpx_file_path}")
    
    try:
        with open(gpx_file_path, "r", encoding="utf-8") as f:
            gpx = gpxpy.parse(f)
    except Exception as e:
        raise ValueError(f"Failed to parse GPX file {gpx_file_path}: {e}")
    
    # Get file size
    file_size = gpx_file_path.stat().st_size
    
    # Collect all points first (without UUIDs)
    records: List[dict] = []
    
    # Process all trackpoints
    for track in gpx.tracks:
        track_name = track.name or None
        for segment in track.segments:
            for point in segment.points:
                if point.time and point.latitude and point.longitude:
                    records.append({
                        SchemaColumns.DATA_TYPE: DataType.GPX_TRACKPOINT,
                        SchemaColumns.NAME: track_name,
                        SchemaColumns.CAPTION: None,
                        SchemaColumns.SOURCE_FILE: str(gpx_file_path),
                        SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: point.time.isoformat(),
                        SchemaColumns.TIMESTAMP_UTC: point.time if point.time.tzinfo is not None else point.time.replace(tzinfo=timezone.utc),
                        SchemaColumns.GPS_LATITUDE: point.latitude,
                        SchemaColumns.GPS_LONGITUDE: point.longitude,
                        SchemaColumns.RATING_0_TO_5: None,
                        SchemaColumns.SIZE_IN_BYTES: file_size,
                    })
    
    # Process all waypoints
    for waypoint in gpx.waypoints:
        if waypoint.time and waypoint.latitude and waypoint.longitude:
            records.append({
                SchemaColumns.DATA_TYPE: DataType.GPX_WAYPOINT,
                SchemaColumns.NAME: waypoint.name,
                SchemaColumns.CAPTION: waypoint.comment or waypoint.description,
                SchemaColumns.SOURCE_FILE: str(gpx_file_path),
                SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: waypoint.time.isoformat(),
                SchemaColumns.TIMESTAMP_UTC: waypoint.time if waypoint.time.tzinfo is not None else waypoint.time.replace(tzinfo=timezone.utc),
                SchemaColumns.GPS_LATITUDE: waypoint.latitude,
                SchemaColumns.GPS_LONGITUDE: waypoint.longitude,
                SchemaColumns.RATING_0_TO_5: None,
                SchemaColumns.SIZE_IN_BYTES: file_size,
            })
    
    if not records:
        logger.warning(f"No valid trackpoints or waypoints found in {gpx_file_path}")
        # Return empty DataFrame with correct schema
        return pd.DataFrame(columns=[
            SchemaColumns.UUID_STRING,
            SchemaColumns.DATA_TYPE,
            SchemaColumns.NAME,
            SchemaColumns.CAPTION,
            SchemaColumns.SOURCE_FILE,
            SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
            SchemaColumns.TIMESTAMP_UTC,
            SchemaColumns.GPS_LATITUDE,
            SchemaColumns.GPS_LONGITUDE,
            SchemaColumns.RATING_0_TO_5,
            SchemaColumns.SIZE_IN_BYTES,
        ])
    
    # Create DataFrame
    df = pd.DataFrame(records)
    
    # Convert TIMESTAMP_UTC to proper datetime with UTC timezone
    df[SchemaColumns.TIMESTAMP_UTC] = pd.to_datetime(df[SchemaColumns.TIMESTAMP_UTC], utc=True)
    
    # Find the first point for UUID generation (first chronologically)
    first_row = df.loc[df[SchemaColumns.TIMESTAMP_UTC].idxmin()]
    first_point_timestamp = first_row[SchemaColumns.TIMESTAMP_UTC].to_pydatetime().replace(tzinfo=timezone.utc)
    first_point_lat = first_row[SchemaColumns.GPS_LATITUDE]
    first_point_lon = first_row[SchemaColumns.GPS_LONGITUDE]
    
    # Generate UUIDs for all rows using vectorized operation
    df[SchemaColumns.UUID_STRING] = df.apply(
        lambda row: generate_uuid_for_row(row, first_point_timestamp, first_point_lat, first_point_lon),
        axis=1
    )
    
    # Set index name
    df.index.name = SchemaColumns.DF_ID
    
    logger.info(f"Loaded {len(df)} points from {gpx_file_path}")
    
    return df