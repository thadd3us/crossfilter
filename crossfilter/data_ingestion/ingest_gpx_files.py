"""CLI program for ingesting GPX files into SQLite database."""

import logging
from pathlib import Path
from typing import List

import pandas as pd
import typer
from sqlalchemy import Column, MetaData, String, Table, create_engine, inspect, text
from sqlalchemy.dialects.sqlite import insert
from sqlalchemy.orm import sessionmaker
from tqdm.contrib.concurrent import process_map
from crossfilter.core.schema import SchemaColumns as C

from crossfilter.core.schema import SchemaColumns
from crossfilter.data_ingestion.gpx_parser import load_gpx_file_to_df

logger = logging.getLogger(__name__)


def find_gpx_files(base_dir: Path) -> List[Path]:
    """Find all GPX files recursively in the base directory."""
    if not base_dir.exists():
        raise FileNotFoundError(f"Base directory not found: {base_dir}")

    if not base_dir.is_dir():
        raise ValueError(f"Base directory is not a directory: {base_dir}")

    gpx_files = list(base_dir.rglob("*.gpx"))
    logger.info(f"Found {len(gpx_files)} GPX files in {base_dir}")

    return gpx_files


def process_single_gpx_file(gpx_file: Path) -> pd.DataFrame:
    """Process a single GPX file and return a DataFrame."""
    try:
        return load_gpx_file_to_df(gpx_file)
    except Exception as e:
        logger.error(f"Failed to process {gpx_file}: {e}")
        # Return empty DataFrame with correct schema on error
        return pd.DataFrame(
            columns=[
                SchemaColumns.UUID_STRING,
                SchemaColumns.DATA_TYPE,
                SchemaColumns.NAME,
                SchemaColumns.CAPTION,
                SchemaColumns.SOURCE_FILE,
                SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
                SchemaColumns.TIMESTAMP_UTC,
                SchemaColumns.GPS_LATITUDE,
                SchemaColumns.GPS_LONGITUDE,
                SchemaColumns.RATING_0_TO_5,
                SchemaColumns.SIZE_IN_BYTES,
            ]
        )


def get_pandas_to_sqlalchemy_dtype(pandas_dtype: str) -> str:
    """Map pandas dtype to SQLAlchemy type string."""
    if pandas_dtype.startswith("int"):
        return "INTEGER"
    elif pandas_dtype.startswith("float"):
        return "REAL"
    elif pandas_dtype.startswith("datetime"):
        return "TIMESTAMP"
    elif pandas_dtype.startswith("bool"):
        return "BOOLEAN"
    else:
        return "TEXT"


def create_or_update_table(engine, table_name: str, df: pd.DataFrame) -> None:
    """Create table or add missing columns if table exists."""
    inspector = inspect(engine)

    if not inspector.has_table(table_name):
        # Create new table
        logger.info(f"Creating new table: {table_name}")
        df.to_sql(table_name, engine, if_exists="replace", index=False)

        # Set UUID_STRING as primary key
        with engine.connect() as conn:
            conn.execute(
                text(
                    f"CREATE UNIQUE INDEX IF NOT EXISTS idx_{table_name}_uuid ON {table_name} ({SchemaColumns.UUID_STRING})"
                )
            )
            conn.commit()
        return True
    else:
        # Check if we need to add columns
        existing_columns = {col["name"] for col in inspector.get_columns(table_name)}
        df_columns = set(df.columns)
        missing_columns = df_columns - existing_columns

        if missing_columns:
            logger.info(f"Adding missing columns to {table_name}: {missing_columns}")

            with engine.connect() as conn:
                for col in missing_columns:
                    dtype = get_pandas_to_sqlalchemy_dtype(str(df[col].dtype))
                    conn.execute(
                        text(f"ALTER TABLE {table_name} ADD COLUMN {col} {dtype}")
                    )
                conn.commit()

        # Check if unique constraint exists on UUID_STRING
        has_unique = has_unique_constraint_on_uuid(engine, table_name)

        if not has_unique:
            # Try to create unique index, but handle the case where duplicates exist
            try:
                with engine.connect() as conn:
                    conn.execute(
                        text(
                            f"CREATE UNIQUE INDEX IF NOT EXISTS idx_{table_name}_uuid ON {table_name} ({SchemaColumns.UUID_STRING})"
                        )
                    )
                    conn.commit()
                logger.info(
                    f"Created unique index on {SchemaColumns.UUID_STRING} for table {table_name}"
                )
                return True
            except Exception as e:
                if "UNIQUE constraint failed" in str(e):
                    logger.warning(
                        f"Cannot create unique index on {table_name}.{SchemaColumns.UUID_STRING} due to existing duplicates. Using fallback upsert strategy."
                    )
                    return False
                else:
                    # Re-raise other errors
                    raise e

        return has_unique


def upsert_dataframe_to_sqlite(
    df: pd.DataFrame, destination_sqlite_db: Path, destination_table: str
) -> None:
    """Upsert DataFrame to SQLite database using ON CONFLICT DO UPDATE or fallback strategy."""
    if df.empty:
        logger.info("No data to upsert")
        return

    # Create database directory if it doesn't exist
    destination_sqlite_db.parent.mkdir(parents=True, exist_ok=True)

    # Create engine
    engine = create_engine(f"sqlite:///{destination_sqlite_db}")

    # Create or update table structure
    create_or_update_table(engine, destination_table, df)

    # Perform upsert using SQLAlchemy
    metadata = MetaData()

    # Reflect the table structure
    table = Table(destination_table, metadata, autoload_with=engine)

    # Convert DataFrame to list of dictionaries
    records = df.to_dict("records")

    # Perform upsert
    with engine.connect() as conn:
        for record in records:
            # Clean None values for SQLite
            cleaned_record = {k: v for k, v in record.items() if v is not None}

            # Use INSERT OR REPLACE for SQLite upsert
            stmt = insert(table).values(cleaned_record)

            # For SQLite, we'll use ON CONFLICT DO UPDATE
            # First, let's check if UUID_STRING is the primary key
            primary_key_col = SchemaColumns.UUID_STRING

            # Create the ON CONFLICT DO UPDATE statement
            update_dict = {
                col.name: stmt.excluded[col.name]
                for col in table.columns
                if col.name != primary_key_col
            }

            if update_dict:
                stmt = stmt.on_conflict_do_update(
                    index_elements=[primary_key_col], set_=update_dict
                )
            else:
                stmt = stmt.on_conflict_do_nothing(index_elements=[primary_key_col])

            conn.execute(stmt)

        conn.commit()

    logger.info(
        f"Upserted {len(records)} records to {destination_table} in {destination_sqlite_db}"
    )


def main(
    base_dir: Path = typer.Argument(
        ..., help="Directory to recursively search for GPX files"
    ),
    destination_sqlite_db: Path = typer.Argument(
        ..., help="SQLite database file to upsert into"
    ),
    destination_table: str = typer.Option("data", help="Database table name"),
    max_workers: int = typer.Option(None, help="Maximum number of parallel workers"),
) -> None:
    """Ingest GPX files into SQLite database."""
    logging.basicConfig(level=logging.INFO)

    # Find all GPX files
    gpx_files = find_gpx_files(base_dir)

    if not gpx_files:
        logger.info("No GPX files found")
        return

    # Process files in parallel
    logger.info(f"Processing {len(gpx_files)} GPX files...")
    dataframes = process_map(
        process_single_gpx_file,
        gpx_files,
        max_workers=max_workers,
        desc="Processing GPX files",
        chunksize=1,
    )

    # Filter out empty DataFrames
    non_empty_dataframes = [df for df in dataframes if not df.empty]

    if not non_empty_dataframes:
        logger.info("No valid data found in GPX files")
        return

    # Concatenate all DataFrames
    logger.info(f"Concatenating {len(non_empty_dataframes)} DataFrames...")
    combined_df = pd.concat(non_empty_dataframes, ignore_index=True)

    combined_df = combined_df.drop_duplicates()

    if combined_df[C.UUID_STRING].duplicated().any():
        logger.warning("Duplicate UUIDs found in data")
        combined_df = combined_df.drop_duplicates(subset=[C.UUID_STRING])

    logger.info(f"Total records: {len(combined_df)}")

    # Upsert to database
    upsert_dataframe_to_sqlite(combined_df, destination_sqlite_db, destination_table)

    logger.info("GPX ingestion completed successfully")


if __name__ == "__main__":
    typer.run(main)
