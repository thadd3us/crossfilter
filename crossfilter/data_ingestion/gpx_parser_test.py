"""Tests for GPX file parsing functions."""

import uuid
from datetime import datetime
from pathlib import Path

from syrupy.assertion import SnapshotAssertion
import pytest

from crossfilter.core.schema import DataType, SchemaColumns
from crossfilter.data_ingestion.gpx_parser import (
    generate_uuid_from_components,
    load_gpx_file_to_df,
)


def test_generate_uuid_from_components() -> None:
    """Test UUID generation from components."""
    timestamp1 = datetime(2023, 1, 1, 12, 0, 0)
    timestamp2 = datetime(2023, 1, 1, 12, 5, 0)

    # Same inputs should generate the same UUID
    uuid1 = generate_uuid_from_components(
        timestamp1,
        37.7749,
        -122.4194,
        timestamp2,
        37.7750,
        -122.4195,
        DataType.GPX_TRACKPOINT,
    )
    uuid2 = generate_uuid_from_components(
        timestamp1,
        37.7749,
        -122.4194,
        timestamp2,
        37.7750,
        -122.4195,
        DataType.GPX_TRACKPOINT,
    )

    assert uuid1 == uuid2
    assert uuid.UUID(uuid1)  # Should be valid UUID

    # Different inputs should generate different UUIDs
    uuid3 = generate_uuid_from_components(
        timestamp1,
        37.7749,
        -122.4194,
        timestamp2,
        37.7750,
        -122.4195,
        DataType.GPX_WAYPOINT,
    )

    assert uuid1 != uuid3


def test_load_gpx_file_to_df_nonexistent_file() -> None:
    """Test loading a non-existent GPX file."""
    nonexistent_file = Path("/nonexistent/file.gpx")

    with pytest.raises(FileNotFoundError):
        load_gpx_file_to_df(nonexistent_file)


def test_load_gpx_file_to_df_valid_file(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test loading a valid GPX file."""
    # Create a simple GPX file
    gpx_content = """<?xml version="1.0" encoding="UTF-8"?>
<gpx version="1.1" creator="test">
    <trk>
        <name>Test Track</name>
        <trkseg>
            <trkpt lat="37.7749" lon="-122.4194">
                <time>2023-01-01T12:00:00Z</time>
            </trkpt>
            <trkpt lat="37.7750" lon="-122.4195">
                <time>2023-01-01T12:05:00Z</time>
            </trkpt>
        </trkseg>
    </trk>
    <wpt lat="37.7751" lon="-122.4196">
        <name>Test Waypoint</name>
        <desc>A test waypoint</desc>
        <time>2023-01-01T12:10:00Z</time>
    </wpt>
</gpx>"""

    gpx_file = tmp_path / "test.gpx"
    gpx_file.write_text(gpx_content)

    df = load_gpx_file_to_df(gpx_file)
    assert list(zip(df.columns, df.dtypes)) == snapshot
    assert df.to_dict(orient="records") == snapshot


def test_load_gpx_file_to_df_empty_file(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test loading an empty GPX file."""
    # Create an empty GPX file
    gpx_content = """<?xml version="1.0" encoding="UTF-8"?>
<gpx version="1.1" creator="test">
</gpx>"""

    gpx_file = tmp_path / "empty.gpx"
    gpx_file.write_text(gpx_content)

    df = load_gpx_file_to_df(gpx_file)
    assert list(zip(df.columns, df.dtypes)) == snapshot
    assert df.to_dict(orient="records") == snapshot


def test_load_gpx_file_to_df_invalid_file(tmp_path: Path) -> None:
    """Test loading an invalid GPX file."""
    invalid_file = tmp_path / "invalid.gpx"
    invalid_file.write_text("This is not a valid GPX file")

    with pytest.raises(ValueError, match="Failed to parse GPX file"):
        load_gpx_file_to_df(invalid_file)


def test_load_gpx_file_to_df_trackpoints_without_time(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test loading a GPX file with trackpoints without timestamps."""
    # Create a GPX file with trackpoints but no timestamps
    gpx_content = """<?xml version="1.0" encoding="UTF-8"?>
<gpx version="1.1" creator="test">
    <trk>
        <name>Test Track</name>
        <trkseg>
            <trkpt lat="37.7749" lon="-122.4194">
            </trkpt>
            <trkpt lat="37.7750" lon="-122.4195">
            </trkpt>
        </trkseg>
    </trk>
</gpx>"""

    gpx_file = tmp_path / "no_time.gpx"
    gpx_file.write_text(gpx_content)

    df = load_gpx_file_to_df(gpx_file)
    assert list(zip(df.columns, df.dtypes)) == snapshot
    assert df.to_dict(orient="records") == snapshot
