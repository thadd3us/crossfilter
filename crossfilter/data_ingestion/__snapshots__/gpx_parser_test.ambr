# serializer version: 1
# name: test_load_gpx_file_to_df_empty_file
  list([
    tuple(
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.NAME: 'NAME'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>,
      datetime64[ns, UTC],
    ),
    tuple(
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>,
      dtype('float64'),
    ),
    tuple(
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>,
      dtype('float64'),
    ),
  ])
# ---
# name: test_load_gpx_file_to_df_empty_file.1
  list([
  ])
# ---
# name: test_load_gpx_file_to_df_trackpoints_without_time
  list([
    tuple(
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.NAME: 'NAME'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>,
      datetime64[ns, UTC],
    ),
    tuple(
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>,
      dtype('float64'),
    ),
    tuple(
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>,
      dtype('float64'),
    ),
  ])
# ---
# name: test_load_gpx_file_to_df_trackpoints_without_time.1
  list([
  ])
# ---
# name: test_load_gpx_file_to_df_valid_file
  list([
    tuple(
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>,
      datetime64[ns, UTC],
    ),
    tuple(
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>,
      dtype('float64'),
    ),
    tuple(
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>,
      dtype('float64'),
    ),
    tuple(
      <SchemaColumns.NAME: 'NAME'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.CAPTION: 'CAPTION'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>,
      dtype('O'),
    ),
  ])
# ---
# name: test_load_gpx_file_to_df_valid_file.1
  list([
    dict({
      <SchemaColumns.CAPTION: 'CAPTION'>: nan,
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>: <DataType.GPX_TRACKPOINT: 'GPX_TRACKPOINT'>,
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>: 37.7749,
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>: -122.4194,
      <SchemaColumns.NAME: 'NAME'>: nan,
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>: 'test.gpx',
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2023-01-01T12:00:00+00:00',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: Timestamp('2023-01-01 12:00:00+0000', tz='UTC'),
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>: 'f50ca1b8-4ef2-5277-f204-ead8ff90010b',
    }),
    dict({
      <SchemaColumns.CAPTION: 'CAPTION'>: nan,
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>: <DataType.GPX_TRACKPOINT: 'GPX_TRACKPOINT'>,
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>: 37.775,
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>: -122.4195,
      <SchemaColumns.NAME: 'NAME'>: nan,
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>: 'test.gpx',
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2023-01-01T12:05:00+00:00',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: Timestamp('2023-01-01 12:05:00+0000', tz='UTC'),
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>: 'd8cc04e9-eaf6-faa8-fc15-8e67a8f6918d',
    }),
    dict({
      <SchemaColumns.CAPTION: 'CAPTION'>: 'A test waypoint',
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>: <DataType.GPX_WAYPOINT: 'GPX_WAYPOINT'>,
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>: 37.7751,
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>: -122.4196,
      <SchemaColumns.NAME: 'NAME'>: 'Test Waypoint',
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>: 'test.gpx',
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2023-01-01T12:10:00+00:00',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: Timestamp('2023-01-01 12:10:00+0000', tz='UTC'),
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>: '0b1274cd-b196-2831-b4f6-166e73220eb8',
    }),
  ])
# ---
